### DISCUSSION TITLE
Build times for our design

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#3353] [General]
There is no question here, just a demonstration of the use of the build time summary for a design that has one top level macro and 5 macros at the top level. What is being built here are mock abstracts for the macros at the top level as we are only interested in regression testing the top level on our build servers.

The top level is the slowest in detailed routing and we have 5 macros at the top level.

We are not currently building the macros at the top level in parallel, though we intend to, but it would only cut the running time by 50%.

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/4cf1cdad-6896-4d2f-9ad2-af2944e50761)

This is the command I run to build a summary of all the build times `find -regex ./logs/.*/base -type d -exec python3 util/genElapsedTime.py -d {} \; | sort -hr -k2`

<!DOCTYPE html>

<html>

<body>

Stage | time/s
-- | --
5_2_TritonRoute | 21313
5_2_TritonRoute | 7284
5_2_TritonRoute | 4744
5_2_TritonRoute | 2964
5_2_TritonRoute | 1825
5_2_TritonRoute | 1482
3_3_place_gp | 1274
4_1_cts | 1158
3_3_place_gp | 666
1_1_yosys | 617
4_1_cts | 379
6_report | 332
3_3_place_gp | 312
5_2_TritonRoute | 294
3_5_opendp | 293
1_1_yosys | 287
5_2_TritonRoute | 282
3_3_place_gp | 278
4_1_cts | 273
3_4_resizer | 224
4_1_cts | 203
5_1_fastroute | 198
6_report | 164
5_1_fastroute | 164
1_1_yosys | 160
5_1_fastroute | 142
3_1_place_gp_skip_io | 132
5_1_fastroute | 126
3_4_resizer | 123
6_report | 116
3_5_opendp | 109
... | ...

</body>

</html>

**Comment by maliberty** [comment]
That is average behaviour as drt is costly.  It does scale pretty well with threads so more CPUs helps.  We do have some work to do distributed drt if you want to test it out but it isn't in the flow currently due to the extra setup overhead.

**Comment by maliberty** [comment]
How many threads in your data above?

**Reply by oharboe** [reply]
12, it's on a local somewhat dated workstation, not on our newer servers.

Ref my earlier plot, it looked like there's little gain beyond 32 threads: https://github.com/The-OpenROAD-Project/OpenROAD/discussions/3298

**Comment by oharboe** [comment]
No question or suggestions here, just a datapoint that I thought it might have some value in sharing.

**Comment by maliberty** [comment]
Fwiw we test with 40 cpus and I know Google often uses 96.   We have tested with several hundred in distributed mode.

**Reply by oharboe** [reply]
We use 128 threads, but at 32, the gains taper off.

Is this not what you are observing?

**Reply by maliberty** [reply]
It is somewhat dependent on design size.  Larger designs should see more benefit.

**Reply by oharboe** [reply]
👍

**Reply by QuantamHD** [reply]
Another important factor to mention is that early in DRT you leverage a lot of parallelism, but towards the end you're left with just a few violations that can often be in single workers.
